### PR TITLE
feat: relativistic HMC

### PR DESCRIPTION
**Pull request by xukai92**

### CODE REVIEW

File: src/AdvancedHMC.jl
Comment (yebai) [inline src/AdvancedHMC.jl:38]: Please consider keeping all these types and functions for relativistic HMC in `experimental/relativistic_hmc.jl`, so we have a clear separation of new research-lead code from the stable code.

File: src/hamiltonian.jl
Comment (yebai) [inline src/hamiltonian.jl:41]: Similiarly, let us keep all these new algorithms under `experimental/relativistic_hmc.jl`

File: src/hamiltonian.jl
Comment (yebai) [inline src/hamiltonian.jl:158]: Same here.

File: src/metric.jl
Comment (yebai) [inline src/metric.jl:116]: Same here.

File: test/hamiltonian.jl
Comment (yebai) [inline test/hamiltonian.jl:15]: Let's keep these tests under `experimental/tests/runtests.jl`. We can then add this script to a new CI action.

File: Project.toml
Comment (yebai) [inline Project.toml:7]: I am not sure about this dependency, but don't want to be blocked by this at this stage. We can revisit this later.

File: src/AdvancedHMC.jl
Comment (yebai) [inline src/AdvancedHMC.jl:41]: Let's not export this, by default. 
```suggestion
```

File: test/experimental/relativistic_hmc.jl
Comment (yebai) [inline test/experimental/relativistic_hmc.jl:8]:
```suggestion
    h = Hamiltonian(metric, AdvancedHMC.RelativisticKinetic(1.0, 1.0), f, g)
```

File: test/experimental/relativistic_hmc.jl
Comment (yebai) [inline test/experimental/relativistic_hmc.jl:9]:
```suggestion
    @test h.kinetic isa AdvancedHMC.RelativisticKinetic
```

**Review by yebai — DISMISSED** [review]
Thanks, @xukai92. I left a few addition minor comments below. This PR should be ready to merge once all tests are passed.

**Comment by xukai92** [comment]
Thanks for the review. I just recalled that we were intended to create a new branch for experimental stuff instead of merging it to the main branch. I will create a new PR for this instead and fix the CI.

**Comment by yebai** [comment]
There are pros and cons for maintaining two branches -- BUG fixes and improvements will take a lot longer to propagate to the main branch. We observed that trend on `DynamicPPL`. 

 It's ok to merge this PR to the master branch I think. The new algorithm is already separated from the stable code, which is the main reason for a new branch.

**Comment by xukai92** [comment]
I’m a bit concerned about the new dependency though. It makes no sense for all upstream packages to depend on a package that is not well maintained (last update was like 2 years ago). I could make it optional using Revise.jo. How do you think?

File: src/AdvancedHMC.jl
Comment (yebai) [inline src/AdvancedHMC.jl:40]: How about only loading experimental algorithms manually to avoid any additional dependency? For tests on CI, we can explicitly call `include("experimental/relativistic_hmc.jl")` from the test script. 
```suggestion
```

**Comment by torfjelde** [comment]
IMO a different branch seems like the best option given the introduction of a new dep :confused:

**Comment by xukai92** [comment]
@yebai I second to @torfjelde. Adding that dependency is a bad idea and managing it using Requires.jl is a bit cumbersome too. Since this is mostly for experimental features, I think we could just create a new branch called `experimental`, which we manually sync with the master. We just need to make sure we only introduce code change in the `src/experimental` folder; all other changes should come from merging with master.

**Comment by yebai** [comment]
> @yebai I second to @torfjelde. Adding that dependency is a bad idea and managing it using Requires.jl is a bit cumbersome too. Since this is mostly for experimental features, I think we could just create a new branch called experimental, which we manually sync with the master. We just need to make sure we only introduce code change in the src/experimental folder; all other changes should come from merging with master.

Sure, that sounds good. Perhaps call that branch `hmc-research` instead?

**Comment by xukai92** [comment]
> Sure, that sounds good. Perhaps call that branch `hmc-research` instead?

Sounds good. I will do this.

> EDIT: in that case, let's perhaps separate the changes to Kinect types into a new PR against master.

Do you mean undo the already merged PR (#299) in master?

File: src/AdvancedHMC.jl
Comment (yebai) [inline src/AdvancedHMC.jl:34]: I am referring these changes -- they should be merged into `master`.

**Pull request closed without merging**
